### PR TITLE
Frontend and ssh add pkdata

### PR DESCRIPTION
**Pull request by HURUIZHE**
Signed-off-by: HURUIZHE <553834398@qq.com>

### CODE REVIEW

File: controllers/cluster/applier/applier.go [hunk 1]
```diff
@@ -54,55 +58,155 @@ func (a *Applier) ReconcileCluster(infra *v1.Infra, hosts []v1.Hosts, cluster *c
...
 	log.Println(cmd)
 	err = ssh.CmdAsync(sealos, cmd, "")
 	if err != nil {
-		return fmt.Errorf("execute ssh commend fail:%v", err)
+		return fmt.Errorf("execute ssh commend %s,fail:%v", cmd, err)
+	}
+	return nil
+}
+func scpSingle(ssh ssh.Interface, ip string) error {
+	if err := ssh.CmdAsync(ip, "echo -e \"#! /bin/bash\\nrole=\\$1\\npassword=\\$2\\nimage=\\$3\\nimageArray=(\\${image//,/ })\\nimages=\\\"\\\"\\nfor var in \\${imageArray[@]}\\ndo\\nimages=\\${images}\\${var}\\\" \\\"\\ndone\\nsudo echo -e \\${password}\\\"\\\\\\n\\\"\\${password} | sudo passwd root\\nsudo hostnamectl set-hostname \\${role}\\nsudo sed -i '63c PasswordAuthentication yes' /etc/ssh/sshd_config\\nsudo systemctl restart sshd\\nsudo wget -c https://github.com/labring/sealos/releases/download/v4.1.3/sealos_4.1.3_linux_amd64.tar.gz && sudo tar zxvf sealos_4.1.3_linux_amd64.tar.gz sealos && sudo rm -rf sealos_4.1.3_linux_amd64.tar.gz && sudo chmod +x sealos && sudo mv sealos /usr/bin\\nsudo sealos run \\${images} --single\" > singleDepl.sh"); err != nil {
```
Comment (fanux) [inline controllers/cluster/applier/applier.go:109]:
```
var shellScript = `
echo -e "xxx"
sudo xxx
`
```

File: controllers/infra/drivers/aws/get_instance.go [hunk 2]
```diff
@@ -326,3 +333,25 @@ func getIndex(i types.Instance) (int, error) {
...
+func retryGetInstance(tryTimes int, trySleepTime time.Duration, client *ec2.Client, inputGetInstance *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+retry:
+	for i := 0; i < tryTimes; i++ {
+		result, _ := GetInstances(context.TODO(), client, inputGetInstance)
+		for _, r := range result.Reservations {
+			for j := range r.Instances {
+				if *r.Instances[j].State.Code == 0 || *r.Instances[j].State.Code == 16 {
+					if r.Instances[j].PublicIpAddress == nil {
+						if i == tryTimes-1 {
+							break
```
Comment (fanux) [inline controllers/infra/drivers/aws/get_instance.go:346]: 太多缩进

**Review by fanux — APPROVED** [review]

File: controllers/infra/drivers/aws/get_instance.go (hunk 2 above)
Comment (HURUIZHE) [reply controllers/infra/drivers/aws/get_instance.go]: modified

File: controllers/cluster/applier/applier.go [hunk 3]
```diff
@@ -54,54 +58,173 @@ func (a *Applier) ReconcileCluster(infra *v1.Infra, hosts []v1.Hosts, cluster *c
...
+done
+
+sudo echo -e ${password}"\n"${password} | sudo passwd root
+sudo hostnamectl set-hostname ${role}
+sudo sed -i '63c PasswordAuthentication yes' /etc/ssh/sshd_config
+sudo /sbin/service sshd restart
+sudo wget -c https://github.com/labring/sealos/releases/download/v4.1.3/sealos_4.1.3_linux_amd64.tar.gz && sudo tar zxvf sealos_4.1.3_linux_amd64.tar.gz sealos && sudo rm -rf sealos_4.1.3_linux_amd64.tar.gz && sudo chmod +x sealos && sudo mv sealos /usr/bin
+sudo sealos run ${images} --single
+`
+	if err := ssh.CmdAsync(ip, "echo "+shellScript+" > singleDepl.sh"); err != nil {
```
Comment (fanux) [inline controllers/cluster/applier/applier.go:128]: fmt.Sprintf("echo %s > single.sh", shellScript)

**Review by fanux — APPROVED** [review]

**Pull request merged**